File: perf-event/examples/locality.rs
/*! Observe the L1 data cache hit rate under two different access patterns.

This example measures L1 data cache hit rate while accessing a 40MB
array linearly, and then randomly.

One surprising finding is that, even though the loop accessing the
array is extremely simple, a `dev` build performs seven times as many
reads as a `release` build. Furthermore, in a `dev` build, the L1 data
cache still manages to achieve an 86% hit rate even under the random
access pattern, which should be completely uncacheable.

This suggests that the machine code for a `dev` build generates a lot
of extraneous memory traffic, but the cache is able to cover for most
of it. This would seem to render `dev` builds unsuitable for assessing
cache behavior.

    $ cargo run --quiet --example locality
    linear: misses / reads:  1251734 / 70038297   1.79%
    random: misses / reads: 10015294 / 70451603  14.22%

    $ cargo run --quiet --example locality --release
    linear: misses / reads:  1250392 / 10000713  12.50%
    random: misses / reads:  9998262 / 10011094  99.87%

On some machines, running the same `--release` executable reports more
misses than reads, which should be impossible. If you know why this
occurs, please file an issue at
`https://github.com/jimblandy/perf-event`.

*/

fn main() {
    use std::hint::black_box;

    const SIZE: usize = 10_000_000;

    // Build a vector that `walk` will traverse from start to end.
    let mut vec: Vec<usize> = (1..SIZE).chain(Some(0)).collect();
    measure("linear", || {
        black_box(walk(&vec));
    });

    let mut random = XorShift128Plus::from_seed(1729, 42);
    random.nth(100); // Propagate the 1-bits in the state a bit.

    // Build a vector that `walk` will access randomly.
    //
    // This turns `vec` into a single chain that visits every element.
    // Proof left to the reader.
    vec.clear();
    vec.extend(0..SIZE);
    let mut rest = &mut vec[..];
    while let Some((first, next @ &mut [_, ..])) = rest.split_first_mut() {
        let swap_with = random.next().unwrap() as usize % next.len();
        std::mem::swap(first, &mut next[swap_with]);
        rest = next;
    }

    measure("random", || {
        black_box(walk(&vec));
    });
}

/// Access elements of `indices`, as guided by its contents.
///
/// Treating each element of `indices` as the index of the next
/// element to visit, start with `indices[0]` and follow the path
/// until we get back to `0`.
///
/// Note that the access pattern depends solely on the slice's
/// contents, not on control flow. The caller can produce whatever
/// access pattern it wants by choosing the contents of `indices`
/// appropriately.
///
/// Return the number of steps needed to return to index 0.
fn walk(indices: &[usize]) -> usize {
    let mut count = 0;
    let mut i = 0;
    loop {
        count += 1;
        i = indices[i];
        if i == 0 {
            break;
        }
    }
    count
}

fn measure(label: &str, task: impl FnOnce()) {
    use perf_event::events::{Cache, CacheOp, CacheResult, WhichCache};
    use perf_event::{Builder, Group};

    let mut group = Group::new().expect("creating group is ok");
    let read_counter = Builder::new()
        .group(&mut group)
        .kind(Cache {
            which: WhichCache::L1D,
            operation: CacheOp::READ,
            result: CacheResult::ACCESS,
        })
        .build()
        .expect("building read_counter is ok");
    let read_miss_counter = Builder::new()
        .group(&mut group)
        .kind(Cache {
            which: WhichCache::L1D,
            operation: CacheOp::READ,
            result: CacheResult::MISS,
        })
        .build()
        .expect("building read_miss_counter is ok");
    group.enable().expect("enabling group is ok");
    task();
    group.disable().expect("disabling group is ok");

    let counts = group.read().expect("reading group is ok");
    let reads = counts[&read_counter];
    let read_misses = counts[&read_miss_counter];

    println!(
        "{label}: misses / reads: {read_misses:8} / {reads:8} {:6.2}%",
        (read_misses as f64 / reads as f64) * 100.0,
    );

    if counts.time_enabled() != counts.time_running() {
        println!(
            "time enabled: {}  time running: {}",
            counts.time_enabled(),
            counts.time_running(),
        );
    }
}

/// The [XorShift128+] pseudorandom number generator.
///
/// This implements [`Iterator`], producing pseudorandom `u64` values
/// as items.
///
/// [XorShift128+]: https://en.wikipedia.org/wiki/Xorshift#xorshift+
struct XorShift128Plus {
    state: [u64; 2],
}

impl XorShift128Plus {
    fn from_seed(seed1: u64, seed2: u64) -> Self {
        assert!(seed1 != 0 || seed2 != 0);
        Self {
            state: [seed1, seed2],
        }
    }
}

impl Iterator for XorShift128Plus {
    type Item = u64;

    fn next(&mut self) -> Option<Self::Item> {
        let mut t = self.state[0];
        let s = self.state[1];
        self.state[0] = s;
        t ^= t << 23;
        t ^= t >> 18;
        t ^= s ^ (s >> 5);
        self.state[1] = t;
        Some(t.wrapping_add(s))
    }
}

File: perf-event/examples/locality.c
/* Observe the L1 data cache hit rate under two different access patterns.

   C equivalent of locality.rs, using perf_event_open directly.
   Written by Claude Code.

   Compile with optimization for meaningful results:
       cc -O2 -o locality locality.c && ./locality
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/syscall.h>
#include <linux/perf_event.h>

#define SIZE 10000000

/* XorShift128+ PRNG: https://en.wikipedia.org/wiki/Xorshift#xorshift+ */
typedef struct { uint64_t s[2]; } XorShift128Plus;

static uint64_t xorshift128plus_next(XorShift128Plus *rng) {
    uint64_t t = rng->s[0];
    uint64_t s = rng->s[1];
    rng->s[0] = s;
    t ^= t << 23;
    t ^= t >> 18;
    t ^= s ^ (s >> 5);
    rng->s[1] = t;
    return t + s;
}

static long perf_event_open(struct perf_event_attr *attr, pid_t pid,
                            int cpu, int group_fd, unsigned long flags) {
    return syscall(__NR_perf_event_open, attr, pid, cpu, group_fd, flags);
}

/* Follow the chain encoded in indices[], starting at 0, until we return to 0.
   Returns the number of steps taken. */
__attribute__((noinline))
static size_t walk(size_t *indices) {
    size_t count = 0;
    size_t i = 0;
    do {
        count++;
        i = indices[i];
    } while (i != 0);
    return count;
}

static int open_cache_counter(int group_fd, int which, int op, int result) {
    struct perf_event_attr attr = { .size = sizeof(attr) };
    attr.type   = PERF_TYPE_HW_CACHE;
    attr.config = (uint64_t)which | ((uint64_t)op << 8) | ((uint64_t)result << 16);

    attr.exclude_kernel = 1;
    attr.exclude_hv     = 1;

    if (group_fd == -1) {
        /* Group leader: start disabled, and carry time/group info on read. */
        attr.disabled    = 1;
        attr.read_format = PERF_FORMAT_GROUP
                         | PERF_FORMAT_TOTAL_TIME_ENABLED
                         | PERF_FORMAT_TOTAL_TIME_RUNNING;
    }

    int fd = (int)perf_event_open(&attr, 0 /* this process */, -1 /* any cpu */,
                                  group_fd, 0);
    if (fd < 0) { perror("perf_event_open"); exit(1); }
    return fd;
}

static void measure(const char *label, size_t *vec) {
    int leader_fd   = open_cache_counter(-1,
                          PERF_COUNT_HW_CACHE_L1D,
                          PERF_COUNT_HW_CACHE_OP_READ,
                          PERF_COUNT_HW_CACHE_RESULT_ACCESS);
    int miss_fd     = open_cache_counter(leader_fd,
                          PERF_COUNT_HW_CACHE_L1D,
                          PERF_COUNT_HW_CACHE_OP_READ,
                          PERF_COUNT_HW_CACHE_RESULT_MISS);
    int prefetch_fd = open_cache_counter(leader_fd,
                          PERF_COUNT_HW_CACHE_L1D,
                          PERF_COUNT_HW_CACHE_OP_PREFETCH,
                          PERF_COUNT_HW_CACHE_RESULT_ACCESS);

    ioctl(leader_fd, PERF_EVENT_IOC_RESET,   PERF_IOC_FLAG_GROUP);
    ioctl(leader_fd, PERF_EVENT_IOC_ENABLE,  PERF_IOC_FLAG_GROUP);

    volatile size_t result = walk(vec);
    (void)result;

    ioctl(leader_fd, PERF_EVENT_IOC_DISABLE, PERF_IOC_FLAG_GROUP);

    /* PERF_FORMAT_GROUP read layout:
         uint64_t nr;            (number of counters = 3)
         uint64_t time_enabled;
         uint64_t time_running;
         uint64_t values[3];     (reads, misses, prefetches)  */
    struct {
        uint64_t nr;
        uint64_t time_enabled;
        uint64_t time_running;
        uint64_t values[3];
    } counts;

    if (read(leader_fd, &counts, sizeof(counts)) != (ssize_t)sizeof(counts)) {
        perror("read"); exit(1);
    }

    uint64_t reads      = counts.values[0];
    uint64_t misses     = counts.values[1];
    uint64_t prefetches = counts.values[2];
    uint64_t hits       = reads - misses;

    printf("%s: hits / reads: %8lu / %8lu  %6.2f%%, prefetched %8lu\n",
           label, (unsigned long)hits, (unsigned long)reads,
           (double)hits / (double)reads * 100.0,
           (unsigned long)prefetches);

    if (counts.time_enabled != counts.time_running)
        printf("time enabled: %lu  time running: %lu\n",
               (unsigned long)counts.time_enabled,
               (unsigned long)counts.time_running);

    close(leader_fd);
    close(miss_fd);
    close(prefetch_fd);
}

int main(void) {
    size_t *vec = malloc(SIZE * sizeof(size_t));
    if (!vec) { perror("malloc"); return 1; }

    /* Linear chain: vec[0]=1, vec[1]=2, ..., vec[SIZE-2]=SIZE-1, vec[SIZE-1]=0 */
    for (size_t i = 0; i < SIZE - 1; i++)
        vec[i] = i + 1;
    vec[SIZE - 1] = 0;

    measure("linear", vec);

    /* Random chain: Fisher-Yates shuffle starting from identity permutation,
       producing a single cycle that visits every element before returning to 0. */
    XorShift128Plus rng = { .s = {1729, 42} };
    for (int i = 0; i < 100; i++)   /* propagate 1-bits in state */
        xorshift128plus_next(&rng);

    for (size_t i = 0; i < SIZE; i++)
        vec[i] = i;

    for (size_t i = 0; i < SIZE - 1; i++) {
        size_t remaining = SIZE - 1 - i;
        size_t j = i + 1 + xorshift128plus_next(&rng) % remaining;
        size_t tmp = vec[i]; vec[i] = vec[j]; vec[j] = tmp;
    }

    measure("random", vec);

    free(vec);
    return 0;
}
